gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamInstrumentation.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.hadoop.fs.statistics.IOStatistics;
+
+/** A class to update the input stream statistics of {@link GoogleHadoopFSInputStreamBase} */
+public class GoogleHadoopFSInputStreamInstrumentation extends GoogleHadoopFSInputStreamBase {
+
+  // Statistic tracker of the Input stream
+  private final GhfsInputStreamStatistics streamStatistics;
+
+  // IO Statistics of the current input stream
+  private final IOStatistics ioStatistics;
+
+  GoogleHadoopFSInputStreamInstrumentation(
+      GoogleHadoopFileSystemBase ghfs,
+      URI gcsPath,
+      GoogleCloudStorageReadOptions readOptions,
+      FileSystem.Statistics statistics)
+      throws IOException {
+    super(ghfs, gcsPath, readOptions, statistics);
+    this.streamStatistics = ghfs.getInstrumentation().newInputStreamStatistics(statistics);
+    this.ioStatistics = streamStatistics.getIOStatistics();
+  }
+
+  GoogleHadoopFSInputStreamInstrumentation(
+      GoogleHadoopFileSystemBase ghfs, FileInfo fileInfo, FileSystem.Statistics statistics)
+      throws IOException {
+    super(ghfs, fileInfo, statistics);
+    this.streamStatistics = ghfs.getInstrumentation().newInputStreamStatistics(statistics);
+    this.ioStatistics = streamStatistics.getIOStatistics();
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    streamStatistics.readOperationStarted(getPos(), 1);
+    int response = 0;
+    try {
+      response = super.read();
+    } catch (IOException e) {
+      streamStatistics.readException();
+      streamStatistics.readOperationCompleted(1, response);
+    }
+    streamStatistics.bytesRead(1);
+    streamStatistics.readOperationCompleted(1, response);
+    return response;
+  }
+
+  @Override
+  public synchronized int read(byte[] buf, int offset, int length) throws IOException {
+    streamStatistics.readOperationStarted(getPos(), length);
+    int response = super.read(buf, offset, length);
+    if (response > 0) {
+      streamStatistics.bytesRead(response);
+    } else {
+      streamStatistics.readException();
+    }
+    streamStatistics.readOperationCompleted(length, response);
+    return response;
+  }
+
+  @Override
+  public synchronized int read(long position, byte[] buf, int offset, int length)
+      throws IOException {
+    streamStatistics.readOperationStarted(position, length);
+    int response = super.read(position, buf, offset, length);
+    if (response > 0) {
+      streamStatistics.bytesRead(response);
+    } else {
+      streamStatistics.readException();
+    }
+    streamStatistics.readOperationCompleted(length, response);
+    return response;
+  }
+
+  @Override
+  public synchronized void seek(long pos) throws IOException {
+    long curPos = getPos();
+    long diff = pos - curPos;
+    try {
+      super.seek(pos);
+      if (diff > 0) {
+        streamStatistics.seekForwards(diff);
+      } else {
+        streamStatistics.seekBackwards(diff);
+      }
+    } catch (IOException e) {
+      throw e;
+    }
+  }
```

Comment:
We can capture stats prior to calling `super.seek(pos)` and get away without `try-catch` block

Reply:
Updated